Sources/Extensions/Foundation/String.swift

```diff
@@ -59,3 +59,91 @@ public extension String {
+
+        let matchSet = CharacterSet(charactersIn: replacementMap.keys.reduce(into: "") { $0 += String($1) })
+            .union(charactersToBeSkipped ?? CharacterSet())
+
+        var final = ""
+
+        let scanner = Scanner(string: self)
+        scanner.charactersToBeSkipped = charactersToBeSkipped
+
+        while !scanner.isAtEnd {
+
+            // copy everything until finding a character to be replaced or skipped
+            var collector: NSString? = ""
+            if scanner.scanUpToCharacters(from: matchSet, into: &collector), let collector = collector {
+                final.append(collector as String)
+            }
+
+            // exit early if we're already at the end
+            guard !scanner.isAtEnd else { break }
+
+            // skip characters if needed
+            if let charactersToBeSkipped = charactersToBeSkipped {
+                scanner.scanCharacters(from: charactersToBeSkipped, into: nil)
+            }
+
+            // find and replace matching character if needed
+            replacementMap
+                .first { match, _ in scanner.scanString(String(match), into: nil) }
+                .flatMap { _, replacement in final.append(replacement) }
+        }
+
+        return final
+    }
+}
+
+extension String {
+
+    public static let nonBreakingSpace = String(Character.nonBreakingSpace)
+    public static let nonBreakingHyphen = String(Character.nonBreakingHyphen)
+    public static let wordJoiner = String(Character.wordJoiner)
+    public static let emDash = String(Character.emDash)
+    public static let enDash = String(Character.enDash)
+
+    /// Returns a non line breaking version of `self`. Line breaking characters occurrences are replaced with
+    /// corresponding non line breaking variants when existent. Otherwise, word joiner characters are attached to them
+    /// to make them non line breaking.
+    ///
+    /// The character mapping is:
+    ///  - space (" ") -> non breaking space (`U+2028`)
+    ///  - hyphen ("-") -> non breaking hyphen (`U+00A0`)
+    ///  - em dash ("—") -> word joiner (`U+2060`) + em dash + word joiner (`U+2060`)
+    ///  - em dash ("–") -> word joiner (`U+2060`) + en dash + word joiner (`U+2060`)
+    ///  - question mark ("?") -> question mark + word joiner (`U+2060`)
+    ///  - closing brace ("}") -> closing brace + word joiner (`U+2060`)
+    ///
+    /// - Returns: A modified version of the receiver without line breaking characters.
+    public func nonLineBreaking() -> String {
+
+        replacingOccurrencesOfCharacters(
+            in: [
+                " ": String.nonBreakingSpace,
+                "-": String.nonBreakingHyphen,
+                .emDash: String([.wordJoiner, .emDash, .wordJoiner]),
+                .enDash: String([.wordJoiner, .enDash, .wordJoiner]),
+                "?": "?" + .wordJoiner,
+                "}": "}" + .wordJoiner,
+            ],
+            skippingCharactersIn: nil
```

Comment:
I believe we want to skip `.newlines` here.

An alternative approach would be to replace occurences of `"\n"`, `"\r"`, [etc...](https://developer.apple.com/documentation/foundation/characterset/1780325-newlines) with empty strings and we could avoid the whole `skippingCharactersIn` which might not be super-intuitive for a general tool.

Reply:
I explicitly removed `.newlines` from being passed by default (from our private implementation, that is 🤭), precisely because it previously had weird behavior when newline characters were present in the middle of the string. Essentially, it preserved newlines (in this case) if they appeared **before** a match.

That led me to adding these fixes ([fix1](https://github.com/Mindera/Alicerce/pull/227/files/0f7a36ac4d757a72ccb61ce700bf27811c458e41#diff-b69213790c7e203256a581e51d5a09b68d06af47e7c3aa10c7845448a38b7b76R81), [fix2](https://github.com/Mindera/Alicerce/pull/227/files/0f7a36ac4d757a72ccb61ce700bf27811c458e41#diff-b69213790c7e203256a581e51d5a09b68d06af47e7c3aa10c7845448a38b7b76R99-R102)) which correctly ignores skipped characters if they appear in the middle of the text, and makes `skippingCharactersIn` behave as expected from an API perspective:

```swift
let matchSet = CharacterSet(charactersIn: replacementMap.keys.reduce(into: "") { $0 += String($1) })
    .union(charactersToBeSkipped ?? CharacterSet())
```

```swift
// skip characters if needed
if let charactersToBeSkipped = charactersToBeSkipped {
    scanner.scanCharacters(from: charactersToBeSkipped, into: nil)
}
```

However, after adding the fixes, then all newline characters were effectively skipped, but never replaced (we have no rule for that) - meaning words would get glued if on separate lines, for instance. Considering that the purpose of this helper is to avoid text from being cut when not _desired_, I think that stripping explicit newlines from the text is perhaps a different transformation. That's why in the end I decided that newlines would be preserved.

Furthermore, replacing newlines with a non breaking space (which seems like the best candidate), might not always produce the desired/expected output either, because in the same way that gluing words/elements in consecutive lines can be an artifact, so can "separating" them. Do you agree?

**PS**: After doing these tests once again, I noticed that [fix2](https://github.com/Mindera/Alicerce/pull/227/files/0f7a36ac4d757a72ccb61ce700bf27811c458e41#diff-b69213790c7e203256a581e51d5a09b68d06af47e7c3aa10c7845448a38b7b76R99-R102) it not needed, as adding the characters to be skipped to the `matchSet` is enough for them to be skipped (i.e. no need for the explicit `scanner.scanCharacters(from: charactersToBeSkipped, into: nil)` after a match is detected).

Reply:
Fair enough 👌